packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -611,12 +611,29 @@ function dbCollectionGroupKey(
     /* document id */ path.length > 0 ? path[path.length - 1] : ''
   ];
 }
+
 /**
  * Comparator that compares document keys according to the primary key sorting
- * used by the `DbRemoteDocumentDocument` store (by collection path and then
- * document ID).
+ * used by the `DbRemoteDocumentDocument` store (by prefix path, collection id
+ * and then document ID).
+ *
+ * Visible for testing.
  */
-function dbKeyComparator(l: DocumentKey, r: DocumentKey): number {
-  const cmp = l.path.length - r.path.length;
-  return cmp !== 0 ? cmp : DocumentKey.comparator(l, r);
+export function dbKeyComparator(l: DocumentKey, r: DocumentKey): number {
+  const left = l.path.toArray();
+  const right = r.path.toArray();
+
+  let cmp = 0;
+  for (let i = 0; i < left.length - 2 && i < right.length - 2; ++i) {
+    cmp = primitiveComparator(left[i], right[i]);
+    if (cmp) {
+      return cmp;
+    }
+  }
+
+  return (
```

Comment:
`primitiveComparator` returns number, not boolean..how this works is not obvious to me..

Can we make it more explicit?

Reply:
JS compares by "thruhiness" and any non-null number is considered true. We could have something like:

```
let cmp = compare()
if (cmp) return cmp;

cmp = compare()
if (cmp) return cmp;

cmp = compare()
if (cmp) return cmp;
```

That was my original code but it is pretty bloated. Let me know what you prefer.

Reply:
I prefer the bloated way..